// @shopify/polaris/esnext/components/TextField/components/Resizer/Resizer.js
import * as React from 'react';
import EventListener from '../../../EventListener';
import styles from '../../TextField.scss';
export default class Resizer extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.contentNode = null;
        this.minimumLinesNode = null;
        this.handleHeightCheck = () => {
            if (this.contentNode == null || this.minimumLinesNode == null) {
                return;
            }
            const contentHeight = this.contentNode.offsetHeight;
            const minimumHeight = this.setMinimumLinesNode
                ? this.minimumLinesNode.offsetHeight
                : 0;
            const newHeight = Math.max(contentHeight, minimumHeight);
            const { currentHeight, onHeightChange } = this.props;
            if (newHeight !== currentHeight) {
                onHeightChange(newHeight);
            }
        };
        this.setContentNode = (node) => {
            this.contentNode = node;
        };
        this.setMinimumLinesNode = (node) => {
            this.minimumLinesNode = node;
        };
    }
    componentDidMount() {
        this.handleHeightCheck();
        if (process.env.NODE_ENV === 'development') {
            // We need to defer the calculation in development so the
            // styles have time to be injected.
            setTimeout(this.handleHeightCheck, 0);
        }
    }
    componentDidUpdate() {
        this.handleHeightCheck();
    }
    render() {
        const { contents, minimumLines } = this.props;
        const minimumLinesMarkup = minimumLines ? (<div testID="MinimumLines" ref={this.setMinimumLinesNode} className={styles.DummyInput} dangerouslySetInnerHTML={{
            __html: getContentsForMinimumLines(minimumLines),
        }}/>) : null;
        return (<div testID="ResizerWrapper" aria-hidden className={styles.Resizer}>
        <EventListener event="resize" handler={this.handleHeightCheck}/>
        <div testID="ContentsNode" ref={this.setContentNode} className={styles.DummyInput} dangerouslySetInnerHTML={{ __html: getFinalContents(contents) }}/>
        {minimumLinesMarkup}
      </div>);
    }
}
const ENTITIES_TO_REPLACE = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '\n': '<br>',
    '\r': '',
};
const REPLACE_REGEX = new RegExp(`[${Object.keys(ENTITIES_TO_REPLACE).join()}]`, 'g');
function replaceEntity(entity) {
    return ENTITIES_TO_REPLACE[entity];
}
function getContentsForMinimumLines(minimumLines) {
    let content = '';
    for (let line = 0; line < minimumLines; line++) {
        content += '<br>';
    }
    return content;
}
function getFinalContents(contents) {
    return contents
        ? `${contents.replace(REPLACE_REGEX, replaceEntity)}<br>`
        : '<br>';
}
